### PR TITLE
UpdateResource merges existing item with payload

### PR DESCRIPTION
**Pull request by aaron-collier**
Adds:
- Overlays payload on existing resource
- Adds new metadata if not in existingResource
- Updates existing tests

### CODE REVIEW

File: datautils/resource.go [hunk 1]
```diff
@@ -28,6 +28,24 @@ func (d *Resource) GetS(key string) string {
...
+}
+
 func (d *Resource) String() string {
 	return fmt.Sprintf("<Resource id: '%s'>", d.ID())
 }
+
+// HasKey returns a boolean value:
+// false unless the key exists
+func (d *Resource) HasKey(key string) bool {
+	if (*d)[key] == nil {
```
Comment (jcoyne) [inline datautils/resource.go]: Could you do `return (*d)[key] != nil`

File: handlers/update_resource.go [hunk 2]
```diff
@@ -83,3 +72,42 @@ func (d *updateResourceEntry) addToStream(id string) error {
 	}
 	return d.stream.SendMessage(string(message))
 }
+
+func (d *updateResourceEntry) compareAndUpdateResource(id string, newResource datautils.Resource, existingResource *datautils.Resource) error {
```
Comment (jcoyne) [inline handlers/update_resource.go]: This seems like something that should go in `datautils` or `db`, it's not a great match for the `handler` package.

Comment (aaron-collier) [reply handlers/update_resource.go]: Hmm... I'm open to discussion on this for sure. I don't think datautils is right (at least with the current implementation) as it would require dataUtils to be aware of the database. Maybe db, but then should db have that kind of logic? It seems like the bulk of the logic really be in the handler.

Comment (jcoyne) [reply handlers/update_resource.go]: My thinking here is that handler should mainly deal with web-interaction stuff.  E.g. identifying the user, and starting the database update, logging and streaming stuff.  This `compareAndUpdateResource` function is not specific to web interactions and could be used outside the web context, (Say if we had a background worker), as it's dealing only with data level concerns.

Comment (aaron-collier) [reply handlers/update_resource.go]: That makes sense. I need to rework this a bit per this mornings discussion of updating when the version changes anyway, so I'll move some of this logic into dataUtils and the actual send to update will happen from the handler.

File: datautils/compare.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
+package datautils
+
+import "encoding/json"
+
+// CompareResource takes an input Resource to compare the called resource and returns a new
+// resource object with overlayed/updated content
+func (d *Resource) CompareResource(resourceToCompare *Resource) (map[string]interface{}, error) {
```
Comment (jcoyne) [inline datautils/compare.go]: Should this return datautils.Resource rather than `map[string]interface{}`?

File: handlers/update_resource.go [hunk 4]
```diff
@@ -83,3 +72,12 @@ func (d *updateResourceEntry) addToStream(id string) error {
...
+
+	resourceForUpdate, err := newResource.CompareResource(existingResource)
+	for k, v := range resourceForUpdate {
+		log.Printf("[%s] = [%v]", k, v)
```
Comment (jcoyne) [inline handlers/update_resource.go]: Seems like this is just doing comparing and logging here now.  No update is happening, so calling this method "compareAndUpdateResource" is a bit misleading.

Comment (aaron-collier) [reply handlers/update_resource.go]: Sorry, this was a midway change that I pushed up while debugging something else, it will be complete later this morning.

Comment (jcoyne) [reply handlers/update_resource.go]: @aaron-collier That's fine, would you mind labeling the PR as WIP?

**Comment by aaron-collier** [comment]
Tagging this WIP until I can get back to verify the tests.

**Comment by aaron-collier** [comment]
Again, marking as WIP while working through updates based on revised goals from 4/2 planning.

File: handlers/update_resource.go [hunk 5]
```diff
@@ -27,45 +24,49 @@ type updateResourceEntry struct {
...
 		}
 		panic(err)
 	}
 
-	if err = d.database.Update(resource); err != nil {
-		panic(err)
-	}
+	newResource = datautils.NewResource(d.mergeJSON(&existingResource.JSON, &newResource.JSON)).WithID(id)
 
-	if err = d.addToStream(id); err != nil {
```
Comment (jcoyne) [inline handlers/update_resource.go]: Are you intentionally removing this call to `addToStream`? That seems unrelated to the stated purpose of this PR.

Comment (aaron-collier) [reply handlers/update_resource.go]: Yes, intentionally as we're removing it as needed. I can back that change out and include it in a follow on PR later though if that's more comfortable. I was in the midst of editing this when we made that call, so I (perhaps aggressively) removed it...

Comment (jcoyne) [reply handlers/update_resource.go]: It seems like if you're going to remove it you'd also want to remove https://github.com/sul-dlss-labs/taco/blob/cd4753007d2eb3e40ce8e27586f7e67f1d3dd5e0/handlers/update_resource.go#L20 and update the NewUpdateResource not to take stream as an argument as well as update the `handlers/server.go` not to pass the argument, etc.  In short, yeah, I think reverting this change here is the easier path.

**Comment by cmh2166** [comment]
I'd like to add some complex / heavily nested tests just to be super sure, but looks g2g for a start.
